pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -18,33 +18,101 @@ package machinesync
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
+	configv1 "github.com/openshift/api/config/v1"
 	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	"github.com/openshift/cluster-capi-operator/pkg/util"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	awscapiv1beta1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta1"
+	capiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+const (
+	capiNamespace string = "openshift-cluster-api"
+	mapiNamespace string = "openshift-machine-api"
+)
+
+var (
+	// errPlatformNotSupported is returned when the platform is not supported.
+	errPlatformNotSupported = errors.New(`error determining InfrastructureMachine
+	type, platform not supported`)
+)
+
 // MachineSyncReconciler reconciles CAPI and MAPI machines.
 type MachineSyncReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+
+	Platform      configv1.PlatformType
+	CAPINamespace string
+	MAPINamespace string
 }
 
 // SetupWithManager sets the CoreClusterReconciler controller up with the given manager.
 func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	infrastructureMachine, infrastructurePredicate, err := getInfrastructureMachineAndPredicateFromProvider(r.Platform)
+	if err != nil {
+		return fmt.Errorf("failed to get InfrastructureMachine from Provider: %w", err)
+	}
+
+	// Allow the namespaces to be set externally for test purposes, when not set,
+	// default to the production namespaces.
+	if len(r.CAPINamespace) == 0 {
+		r.CAPINamespace = capiNamespace
+	}
+
+	if len(r.MAPINamespace) == 0 {
+		r.MAPINamespace = mapiNamespace
+	}
```

Comment:
Any particular reason for using `len() == 0` instead of ` == ""` for these strings?

Reply:
Oh.. I thought I'd learnt somewhere along the way that comparing the length was better, although googling now it seems that may be wrong.. will change for  `== ""`!

Reply:
I've also heard that length is better but have no strong opinion, I normally go for compare to empty string personally